### DISCUSSION TITLE
Nasazení OntoGrapheru mimo Výrobní linku

### DISCUSSION THREAD

**Discussion by bindeali** [opendata-mvcr/ontoGrapher#493] [General]
Pro zprovoznění OntoGrapheru je vyloženě nutná pouze grafová databáze. Silně doporučuji mít ale nastavenou i OIDC službu a SGOV server.

## Proměnné prostředí
> ⚠ Pro každou proměnnou je potřeba přidat předponu `REACT_APP_[proměnná]`. ⚠

Pro nasazení jsou nutné proměnné `COMPONENTS`, `DEBUG`, `AUTHENTICATION`.

### DEBUG (default: "false")
Pomocné funkce pro vývojové/netlify prostředí, kde není přístup ke grafové databázi. Načte nastavení z `./src/debug-data.json`. Tyto data jdou v OntoGrapheru kdykoliv vygenerovat pomocí `Ctrl+Alt+D`. Neukládají se žádné změny.

### AUTHENTICATION (default: "true")
Povoluje/zakazuje OIDC. V případě hodnoty `false` nejsou dostupné funkce přihlášení/odhlášení, validace, sledování změn. Doporučuji nechat na `true` mimo lokální prostředí, jinak je možné, že přestane fungovat komunikace s DB.

### COMPONENTS
Do base64 zakódovaný obsah [tohoto souboru](https://github.com/datagov-cz/sgov-assembly-line/blob/main/components.yml). Pro OntoGrapher je nutné mít nastaveno

* al-db-server:url - URL repositáře grafové databáze podporující RDF4J REST API

Další hodnoty nejsou vyloženě nutné, ale silně je doporučuji nastavit:
* al-auth-server:url - URL OIDC služby
* al-sgov-server:url - URL SGOV serveru, nutné pro validaci (není ale potřeba při AUTHENTICATION="false")
* al-issue-tracker-meta - URL pro tlačítka "Nahlásit chybu" a "Navrhnout úpravu"

## Práce v OntoGrapheru
Vzhledem k tomu, že není nasazen Kontrolní panel, tak je potřeba předem znát URL slovníkových kontextů, které chceme používat. Tento `vocabulary` parametr je nutný zadat; zatím chybí podpora načtení všech slovníkových kontextů v případě nedodání parametru.

V praxi OntoGrapher nerozlišuje mezi kontexty, které mu předává Kontrolní panel a těmi, tak, jak jsou nyní v [Cache](https://xn--slovnk-test-scb.mvcr.gov.cz/modelujeme/sluzby/db-server/resource?uri=https:%2F%2Fslovn%C3%ADk.gov.cz) . Stačí, aby v předávaném kontextu měl slovník 
* rdf:type http://onto.fel.cvut.cz/ontologies/slovník/agendový/popis-dat/pojem/slovník,
* http://purl.org/dc/terms/title `název`,
* http://onto.fel.cvut.cz/ontologies/slovník/agendový/popis-dat/pojem/má-glosář `glosář`.

OntoGrapher ale používá Cache pro hledání pojmů a slovníků "mimo projekt", a tak souběžné upravování Cache pojmů jako "projektových" slovníků a přidávání jako "read-only" slovníků není nijak ošetřené.

### Co v standalone verzi chybí (podle mého názoru)
* Lepší management diagramů. V nynější verzi se každý diagram ukládá ke každému projektovému slovníku, a tak práce s více slovníky s sebou přináší zmatek a velké nároky na bandwith. Upravila bych to, aby se každý nový diagram začínal úplně od začátku, tj. s žádnými slovníky - postupně by si uživatel přidával ty, se kterými chce v daném diagramu pracovat.
* Otevření všech slovníků při nedodání `vocabulary` parametru.

**Comment by bindeali** [comment]
### Návrh standalone verze
* V případě nedodání `vocabulary` OG nalezne a načte **_všechny_** slovníky (duplikáty buď musí OG vyfiltrovat, nebo se musí z DB odstranit).
* S diagramy se pracuje tak, že si nové pojmy/slovníky uživatel vyhledává stejně, jako by si je hledal v cache, akorát všechny pojmy a slovníky mají povolený zápis (read-only cache tedy neexistuje). 
* Pro každého uživatele se ukládá množina diagramů, které má právě otevřené - začíná se se všemi zavřenými.

Velmi jednoduše se mohou stát konflikty, kdy někdo někomu mění slovník "pod rukama". Ideální řešení je server. Jednoduché, ale problematické řešení je nějaký lock přes DB.

**Comment by bindeali** [comment]
## Aktualizace:
Branch `standalone-basic` má minimální implementaci standalone verze. Je přidána funkce načítající všechny slovníky, pokud je nastavená proměnná `STANDALONE=true` a není dodán URL parametr `vocabulary`. 

Daná verze je, jak již bylo řečeno, minimální. Nejsou tam (podle mě) zásadní optimalizace pro standalone práci, například:
* HW nároky na server i klienta s počtem slovníků v DB roste v podstatě exponenciálně. 
* Žádný lock/podpora souběžné spolupráce.

Návrh kompletního řešení je v [tomto komentáři](https://github.com/datagov-cz/ontoGrapher/discussions/493#discussioncomment-5873493).

**Comment by MichalMed** [comment]
Načítá OG standardně nějaké externí slovníky, které potřebuje pro základní funkčnost? Při standalone nasazení nejsou k dispozici jiné vztahy než je podtřídou (chybí např. má vtzažený prvek 1/2 atd..).
